Dialog vs popup (#735)
JetBrains/kotlin-multiplatform-dev-docs: topics/compose-popups.md

[//]: # (title: Popups)

> A `Popup()` is a floating container that renders its content on top of the current UI.

Reviewer: Should we add something about `Dialog()` (common API) here to actually describe the difference (the initial request from YT)?

Author: I have this in the next sentence:

> Unlike a modal dialog, a popup is non-modal, has no scrim, and is not centered by default, 
> which makes it suitable for dropdowns, tooltips, and custom overlays anchored to a component.

Probably can make it more visible

Reviewer: > a modal dialog

There is no reference to a specific API. So, I see it as `Dialog()` API is not documented

Author: I added an inline definition of a dialog on first mention, plus a short guidance paragraph explaining when to use a dialog versus a popup

[//]: # (title: Popups)

> A `Popup()` is a floating container that renders its content on top of the current UI within the same window.

Reviewer 2: ```suggestion
A popup in Compose Multiplatform is a floating container that renders its content on top of the current UI within the same window.
```
To not confuse the scopes: a function is not a container.

> Unlike the multiplatform `Dialog()` API — which acts as a modal container that takes focus, centers its content, 

Reviewer 2: ```suggestion
Unlike the multiplatform `Dialog()` API – which acts as a modal container that takes focus, centers its content, 
```

> and uses a dimmed scrim to block interaction with the rest of the UI — a `Popup()` is non-modal. 

Reviewer 2: ```suggestion
and uses a dimmed scrim to block interaction with the rest of the UI – a `Popup()` is non-modal. 
```

> and uses a dimmed scrim to block interaction with the rest of the UI — a `Popup()` is non-modal. 

Reviewer 2: I'm just fixing the dashes, but the part between the dashes is too big relative to the rest. A simpler approach would be "Dialog acts as a modal container ... Popup, on the other hand, is non-modal."

It has no scrim, does not restrict its width, and allows users to continue interacting with the underlying UI.

> A popup is not centered by default and requires explicit positioning to anchor it to a component.

Reviewer 2: ```suggestion
A popup is not centered by default and requires additional arguments to anchor it to a component.
```
"Positioning" may be too general.

> Use `Dialog()` when you need to interrupt the user and require a decision before they continue, 

Reviewer 2: We have a link for `DialogWindow()`, but not for `Dialog()`, and no samples for `Dialog()` either. Will it be a separate page in the future? Or can we link to DAC straight away?

such as confirmations, alerts, or short forms. For separate OS-level dialogs on desktop, 
see [`DialogWindow()`](compose-desktop-top-level-windows-management.md#dialogs).
Use `Popup()` for lightweight, non-blocking overlays that stay anchored to a component inside the current window, 
such as dropdowns, tooltips, and menus.

## Positioning a popup

To position a popup, use either an `alignment` and `offset`, or a custom
`PopupPositionProvider` for anchored placement.

For a simple alignment:

```kotlin
var isPopupOpen by remember { mutableStateOf(false) }

Box(Modifier.padding(24.dp)) {
    Button(onClick = { isPopupOpen = !isPopupOpen }) {
        Text("Toggle popup")
    }

    if (isPopupOpen) {
        Popup(
            // Positions the popup relative to the button
            alignment = Alignment.TopStart,

>             offset = IntOffset(0, 120),

Reviewer 2: ```suggestion
            offset = IntOffset(30, 70),
```

Reviewer 2: This and the other code suggestion is to make the popup more visibly on top of the rest of the UI, otherwise it just looks like a toggled `Box()`.

Also I'm not sure as I'm reading the code how the offset is implemented — it's not `dp`, is it just pixels? Should we mention this in a comment?

>             // Hides the popup when the user presses outside of it

Reviewer 2: ```suggestion
            // Hides the popup when it is dismissed,
            // for example, when the user clicks outside it
```
Pressing outside the popup is not the only dismiss request as I understand (although if we lose `focusable = true` the Esc method doesn't work).

            onDismissRequest = { isPopupOpen = false },
            // Makes the popup focusable so it can receive key events

>             properties = PopupProperties(focusable = true)

Reviewer 2: This contradicts (or maybe just doesn't illustrate) the statement at the top about "allows users to continue interacting with the underlying UI". As far as I can see when `focusable` is set I can't interact with the button any more.

        ) {
            Box(
                Modifier
                    .background(Color.LightGray, RoundedCornerShape(4.dp))
                    .padding(12.dp)
            ) {

>                 Text("Popup content")

Reviewer 2: ```suggestion
                Text("Popup content on top of UI")
```

            }
        }
    }
}
```

For anchored placement, use a `PopupPositionProvider`:

```kotlin
var isPopupOpen by remember { mutableStateOf(false) }

val belowAnchor = remember {
    object : PopupPositionProvider {
        override fun calculatePosition(
            anchorBounds: IntRect,
            windowSize: IntSize,
            layoutDirection: LayoutDirection,
            popupContentSize: IntSize

>         ) = IntOffset(x = anchorBounds.left, y = anchorBounds.bottom)

Reviewer 2: ```suggestion
        ) = IntOffset(x = anchorBounds.left - 20, y = anchorBounds.bottom - 20)
```
Same as the previous, to make it more popup-y.

    }
}

Column(Modifier.padding(24.dp)) {
    Box {
        Button(onClick = { isPopupOpen = !isPopupOpen }) {
            Text("Toggle menu")
        }

        if (isPopupOpen) {
            Popup(
                popupPositionProvider = belowAnchor,
                onDismissRequest = { isPopupOpen = false },

>                 properties = PopupProperties(focusable = true)

Reviewer 2: Also `focusable = true`, see the other comment about that.

            ) {
                Box(
                    Modifier
                        .shadow(4.dp, RoundedCornerShape(4.dp))
                        .background(Color.White, RoundedCornerShape(4.dp))
                        .padding(12.dp)
                ) {

>                     Text("Anchored below the button")

Reviewer 2: ```suggestion
                    Text("Anchored to the button")
```

                }
            }
        }
    }
}
```

## Customizing behavior

With `PopupProperties`, you can control how the popup handles focus and dismissal:

* `focusable` determines whether the popup receives key events, disabled by default.

>   It must be enabled for `dismissOnBackPress` to work.

Reviewer 2: ```suggestion
```
Stating the requirement where it's required should be enough.

* `dismissOnBackPress` dismisses the popup on Android's back button or the **Esc** key on desktop, 
  enabled by default. Requires `focusable = true`.
* `dismissOnClickOutside` dismisses the popup when the user presses outside its bounds, enabled by default.

`Popup()` and its `PopupProperties` are part of the common API. However,
some properties are only available in platform-specific source sets. For example, on iOS,

> `PopupProperties` provides `usePlatformInsets`, which limits the popup's content to the platform insets (the safe area).

Reviewer 2: This makes it seem like `usePlatformInsets` is only available on iOS when it also works on desktop. Is it more about the fact that you can't use these properties in the common source set?

Author: This property has no effect on desktop. I've updated the wording

> ## What's next

Reviewer 2: ```suggestion
## See also
```

It's not necessarily what's next :)

Author: Let's keep it for consistency

Reviewer 2: I'd say that's not the good kind of consistency :) but not worth focusing on.

[//]: # (title: Popups)

<web-summary>Learn how to create, position, and customize popups in Compose Multiplatform for dropdowns, tooltips, and menus.</web-summary>

A popup in Compose Multiplatform is a floating container that renders its content on top of the current UI within the same window.

Unlike the multiplatform `Dialog()` API, a `Popup()` is non-modal. 
A dialog in Compose Multiplatform acts as a modal container that takes focus, centers its content, 
and uses a dimmed scrim to block interaction with the rest of the UI.
A popup, on the other hand, has no scrim, does not restrict its width, and allows users to continue interacting with the underlying UI.
It is not centered by default and requires additional arguments to anchor it to a component.

Use [`Dialog()`](https://developer.android.com/reference/kotlin/androidx/compose/ui/window/Dialog.composable)

> when you need to interrupt the user and require a decision before they continue, 

Reviewer 2: "decision such as confirmations" is a bit awkward. Maybe "...before they continue, for example, react to confirmations, alerts, or short forms."

        if (isPopupOpen) {
            Popup(
                popupPositionProvider = belowAnchor,
                onDismissRequest = { isPopupOpen = false }
            ) {
                Box(
                    Modifier
                        .shadow(4.dp, RoundedCornerShape(4.dp))
                        .background(Color.White, RoundedCornerShape(4.dp))
                        .padding(12.dp)
                ) {
                    Text("Anchored to the button")
                }
            }
        }
    }
}
```

## Customize behavior

With `PopupProperties`, you can control how the popup handles focus and dismissal:

* `focusable` determines whether the popup receives key events, disabled by default.
* `dismissOnBackPress` dismisses the popup on Android's back button or the **Esc** key on desktop, 
  enabled by default. Requires `focusable = true`.
* `dismissOnClickOutside` dismisses the popup when the user presses outside its bounds, enabled by default.

`Popup()` and its `PopupProperties` are part of the common API. 
However, some properties are not available in the common source set. 

> For example, `usePlatformInsets` is available on iOS and desktop. Its primary use case is on iOS, 

Reviewer 2: If it has no effect on desktop, I wouldn't mention it — it seems more of a me problem, not realizing that desktop has no platform insets to use (even if the API is available there because why not; same for web I guess).
